backend/app/src/main/java/com/moh/phlat/backend/service/SourceDataServiceImpl.java

```diff
@@ -35,81 +36,106 @@ public class SourceDataServiceImpl implements SourceDataService {
 	private SourceDataFilterSpecification specificationService = new SourceDataFilterSpecificationImpl();
 	
 	@Override
-	public List<SourceData> getSourceData(Long controlId, SourceDataFilterParams filterProcess) {
-		
-		Specification<SourceData> combinedSpecification;
+	public List<SourceData> getSourceData(Long controlId, SourceDataFilterParams filterSource, Pageable pageable) {
 		
 		try {
-			combinedSpecification = specificationService.buildSpecificationWhereEqual("controlTableId", controlId.toString());
+			return sourceDataRepository.findAll(buildSpecification(controlId, filterSource), pageable);
 		} catch (Exception e) {
 			logger.error("Error occured: {}", e.getMessage(), e);
 			
 			return new ArrayList();
 		}
+	}
+	
+	@Override
+	public Long countSourceData(Long controlId,  SourceDataFilterParams filterSource) {
+		try{
+			return sourceDataRepository.count(buildSpecification(controlId, filterSource));
+		} catch (Exception e) {
+			logger.error("Error occured: {}", e.getMessage(), e);
+			
+			return null;
+		}
+	}
+	
+	private Specification<SourceData> buildSpecification(Long controlId, SourceDataFilterParams filterSource) throws Exception{
+		
+		Specification<SourceData> combinedSpecification;
+		
+		try {
+			combinedSpecification = specificationService.buildSpecificationWhereEqual("controlTableId", controlId.toString());
+		} catch (Exception e) {
+			throw new Exception(e);
```

Comment:
any reason for catching exception here?

Reply:
Failing early.  This is the only point of failure in building the specification and I figured instead of having all the lines in the try catch, just have the little that will fail.

Reply:
ok, maybe I understood wrong. We can discuss. we even don't need catch block because you are having throws clause in the method. Catch block would have useful if you are doing something like converting it into different exception or returning something else. Which you are doing where this method is called - logging and returning something else.